src/main/java/io/molnarsandor/trelloclone/project/ProjectService.java

```diff
@@ -0,0 +1,141 @@
+        projectEntity.setUser(userEntity);
+        projectEntity.setProjectLeader(userEntity.getEmail());
+        projectEntity.setProjectIdentifier(projectEntity.getProjectIdentifier().toUpperCase());
+
+        if (projectEntity.getId() == null) {
+            BacklogEntity backlogEntity = new BacklogEntity();
+            projectEntity.setBacklog(backlogEntity);
+            backlogEntity.setProject(projectEntity);
+            backlogEntity.setProjectIdentifier(projectEntity.getProjectIdentifier().toUpperCase());
+        }
+
+        if (projectEntity.getId() != null) {
+            projectEntity.setBacklog(backlogRepository.findByProjectIdentifierIgnoreCase(projectEntity.getProjectIdentifier().toUpperCase()));
+        }
+
+        return modelConverter.projectEntityToDto(
+                projectRepository.save(projectEntity));
+
+    }
+
+    public List<ProjectDTO> findAllProject(String username) {
+
+        List<ProjectEntity> byLeader = projectRepository.findAllByProjectLeader(username);
+
+        CollaboratorEntity collaboratorEntity = collaboratorRepository.findByEmail(username);
+
+        List<ProjectEntity> byCollaborator = projectRepository.findAllByCollaborators(collaboratorEntity);
+
+        byLeader.addAll(byCollaborator);
+
+        return modelConverter.projectEntityListToDto(byLeader);
+    }
+
+    public DeleteDTO deleteProjectByIdentifier(String projectId, String username) {
+
+        projectRepository.delete(findProjectByIdentifier(projectId, username));
+
+        return new DeleteDTO("Project " + projectId + " deleted");
+    }
+
+    // == PRIVATE METHODS ==
+    private void validateProject(ProjectEntity projectEntity, String projectIdentifier, String username) {
+        if (projectEntity != null && !projectEntity.getProjectLeader().equals(username)) {
+            throw new ProjectNotFoundException("Project not found in your account");
+        } else if (projectEntity == null) {
+            throw new ProjectNotFoundException("Project with ID: '" + projectIdentifier + "' cannot be updated because it does not exist");
+        }
+    }
+
+    private void validateProjectWithCollaborators(ProjectEntity projectEntity, String projectId, String username) {
+        if (projectEntity == null) {
+            throw new ProjectNotFoundException("Project ID '" + projectId + "' does not exists");
+        }
+
+        Set<CollaboratorEntity> collaboratorEntities = projectEntity.getCollaborators();
+
+        Predicate<CollaboratorEntity> collaboratorPredicate = collaborator ->
+                collaborator.getProjectIdentifier().equalsIgnoreCase(projectId) &&
+                        collaborator.getEmail().equals(username);
+
+
+        boolean collaboratorExists = false;
+        if (!collaboratorEntities.isEmpty()) {
+            collaboratorExists = collaboratorEntities
+                    .stream()
+                    .anyMatch(collaboratorPredicate);
```

Comment:
Ha jól látom, a `projectEntity.collaborators` az NEM `Eager,` emiatt minden egyes elem a `Collection`-ben egyesével lesz felszedve az adatbázisból, tehát pl egy 50 elemű `Collection` esetén ez 50 adatbázis lekérdezést jelent, ami nem túl jó (fölöslegesen terheled az adatbázist, és nem is gyors). Szerintem ezt is egyetlen adatbázis lekérdezéssel meg lehetne oldani az iteráció helyett (kevesebb kód, kevesebb logika, könnyebb megérteni). Ugyanezek vonatkoznak pl. a `CollaboratorService`-ben az ugyanilyen validációkra.